[LangRef] Clarify definition of fragments and debug intrinsics

slinder1 opened the pull request:
Try to give a more complete description of what we call a "fragment", and define "spatial overlap" between variables and fragments to make the details of when one intrinsic ends the effects of another more direct and complete.

Also change the description of the debug intrinsics to the pseudo-grammar style from LangRef.

slinder1 committed:
[LangRef] Clarify definition of fragments and debug intrinsics

Try to give a more complete description of what we call a "fragment",
and define "spatial overlap" between variables and fragments to make
the details of when one intrinsic ends the effects of another more
direct and complete.

Also change the description of the debug intrinsics to the
pseudo-grammar style from LangRef.

felipepiovezan commented on llvm/docs/SourceLevelDebugging.rst:
```diff
@@ -130,6 +130,40 @@ debugging information influences optimization passes then it will be reported
+<LangRef.html#diglobalvariable>`_ metadata nodes.
+
+When variables are not allocated to contiguous memory or to a single LLVM value
+the metadata must record enough information to describe each "piece" or
```
One minor concern here, we say that:

> When variables are not allocated

And then one of two options:

> to contiguous memory
or 
> to a single LLVM value

This second part feels off to me: "allocated to a single LLVM value"

One suggestion would be to avoid mentioning allocation, and instead combine the two possibilities in terms of "describing":

"When variables cannot be described by a single use of an LLVM value, ...".

felipepiovezan replied:
Btw I really like the wording of everything else!

SLTozer replied:
I agree that the wording could be improved - I think that your proposed wording is better, but might be ambiguous with respect to variadic debug values, which aren't necessarily fragments but may be "described" by many LLVM values. Perhaps "When variables are split up across multiple LLVM values...", and it may also be worthwhile to be direct and give an example, i.e. "...such as for structs that have been split into separate LLVM values for each member...". No particularly strong feelings on this part from my end though!

felipepiovezan replied:
Could we say "described by a single expression" then? IIRC when we have multiple fragments, each fragment is described by its own expression (and its own intrinsic)

slinder1 replied:
I agree, the wording I proposed is imprecise/ambiguous; I think I also bled over concepts from the "use fragments for everything" world I was originally thinking in terms of.

I like the framing in terms of the number of expressions for many cases, but we still have cases where we use a single fragment+expression to describe that only one piece of a variable is defined, right? Maybe rather than starting with the "reason" for fragments I can start with just exactly what they are, then in a note go into the various cases they are used and give some examples?

felipepiovezan replied:
>  Maybe rather than starting with the "reason" for fragments 

I think you make a good point here. We don't need to justify the reason for fragments here (arguably it is self evident); we can just state what fragments are:

"A variable may be described in terms of fragments, where a fragment is a contiguous span of bits. This is achieved by ... [rest of the second paragraph]".

If we want to provide motivation, we can do so later through an example or just state in high level terms one code transformation that may cause fragments to appear.

SLTozer reviewed (commented):
Aside from the comments above, this all LGTM - thanks for this!

adrian-prantl commented on llvm/docs/SourceLevelDebugging.rst:
```diff
@@ -130,6 +130,40 @@ debugging information influences optimization passes then it will be reported
+Variables and Variable Fragments
+================================
+
+Source language variables (or just "variables") are represented by `local
```
should we add "and constants"?

SLTozer replied:
Just imo, but I think we could skip mentioning constants, on the basis that the sentence as written isn't exclusive (it doesn't specify that _only_ variables are represented by these classes), and I don't believe(?) that we use fragments for constants since we should always have a single whole definition for them, so the following text would only concern variables. No complaints from me about adding a mention of constants here, though.

slinder1 replied:
I think the terminology gets a bit fuzzy, because "constants" can mean different things in different source languages anyway, right?

I didn't audit every other mention of "variable" in the document, but I imagine this same ambiguity comes up elsewhere, so it would be ideal to get a consistent definition to point back at. DWARF seems to call these "data objects" and says:

```
Chapter 4
Data Object and Object List Entries
This section presents the debugging information entries that describe individual
data objects: variables, parameters and constants, and lists of those objects that
may be grouped in a single declaration, such as a common block.
```

Maybe we can just adopt "data object" here?

SLTozer replied:
In the interest of not bike-shedding this patch forever, I'll avoid commenting too much more (it already looks "good enough" to me!), but I think using "data object" outside of the DWARF context may just be adding confusion. _Personally_ I feel like "Variable" is a good term, since the information here explicitly applies to just "variables" and not really to constants, but ultimately I put my vote for either "variables" or "variables and constants".

adrian-prantl commented on llvm/docs/SourceLevelDebugging.rst:
```diff
@@ -417,8 +448,9 @@ values through compilation, when objects are promoted to SSA values an
 
-* A dbg.value terminates the effect of any preceding dbg.values for (any
-  overlapping fragments of) the specified variable.
+* A dbg.value terminates the effect of any preceding dbg.values for any
```
should we be consistent and use `llvm.dbg.value` here?

adrian-prantl commented on llvm/docs/SourceLevelDebugging.rst:
```diff
@@ -417,8 +448,9 @@ values through compilation, when objects are promoted to SSA values an
-* A dbg.value terminates the effect of any preceding dbg.values for (any
-  overlapping fragments of) the specified variable.
+* A dbg.value terminates the effect of any preceding dbg.values for any
+  spatially overlapping variables or fragments of the specified variable or
```
The way this is worded sounds as if in this example:
```
%1 = ...
call llvm.dbg.value(metadata %1, DILocalVariable("x"), DIExpression())
%2 = ...
call llvm.dbg.value(metadata %1, DILocalVariable("y"), DIExpression())
```
The second intrinsic ends the range of "x", but that's not the case. It only terminates it if it refers to the _same_ variable or a fragment of it.

SLTozer replied:
```suggestion
  spatially overlapping llvm.dbg.values of the specified variable or
```
Perhaps this - I think given the definition of spatial overlaps, it should be clear that it applies to both fragment and non-fragment variable definitions.

slinder1 replied:
Right, it might be good to add a note here, and to make the term "spatially overlapping" be a link back to the definition, but the intention was to define "spatial overlap" such that this could just talk in terms of the universe of all variables and fragments and by definition still be restricted to the way LLVM currently works. In other words, a `dbg.value` for "x" cannot spatially overlap with a `dbg.value` for "y" by definition, so the wording here does not need to repeat that restriction.

slinder1 committed:
Reword and refine definitions

I tried to include more kinds of objects under the umbrella term
"variable", but I'm not particularly happy with how circular it sounds.
Any more thoughts on better approaches are appreciated.

I also didn't understand initially that the current behavior in the most
common cases of spatial overlap between fragments is not necessarily
the intended semantics. Instead it is just a legal shortcut
that balances the complexity of handling partial overlap with arbitrary
expressions with the utility to the debugger for these rarer occurences.

None of this is terribly exact, but I still think having something
written about these aspects is better than having nothing, and might
save future contributors a bit of the learning curve. Let me know what
you think!

OCHyams reviewed (commented):
> I tried to include more kinds of objects under the umbrella term
"variable", but I'm not particularly happy with how circular it sounds.
Any more thoughts on better approaches are appreciated.

IMO what you have now is good.

This all LGTM, thank you for working on this!

I think everyone's previous concerns have been addressed, but I'll hold off on Accepting to give the other reviewers a chance to have another look at this.

jryans reviewed (approved):
Thanks for working on this, seems like a good improvement to me. 😄

slinder1 commented:
I forgot to land this, and enough time has passed that it needed some rework, so I posted #145582 instead
